src/wallet/spend.cpp

```diff
@@ -102,15 +102,19 @@ void CoinsResult::Clear() {
     coins.clear();
 }
 
-void CoinsResult::Erase(std::set<COutPoint>& preset_coins)
+void CoinsResult::Erase(const std::set<COutPoint>& outs)
 {
-    for (auto& it : coins) {
-        auto& vec = it.second;
-        auto i = std::find_if(vec.begin(), vec.end(), [&](const COutput &c) { return preset_coins.count(c.outpoint);});
-        if (i != vec.end()) {
-            vec.erase(i);
-            break;
-        }
+    std::set<COutPoint> coins_to_remove = outs;
+    for (auto& [type, vec] : coins) {
+        auto remove_it = std::remove_if(vec.begin(), vec.end(), [&](const COutput& coin) {
+            auto it = coins_to_remove.find(coin.outpoint);
+            if (it == coins_to_remove.end()) return false;
+
+            // remove coin
+            coins_to_remove.erase(it);
```

Comment:
In 7f6abb39159cd55b6c9193630ca4c3d9ae551be6 "wallet: bugfix, 'CoinsResult::Erase' is erasing only one output of the set"

How come we need to have `coins_to_remove` and erase the removed coins from it? `coins_to_remove` isn't being used anywhere else so this doesn't seem to be doing anything?

Reply:
It's to decrease the search time while moving across the `CoinsResult` vectors.

`Erase` receives a const set with all the coins that needs to be removed from the `CoinsResult` internal vectors. So, what we do is loop across each coins vector and, for each coin inside them, **we test whether the coin is inside the set of coins provided by the caller**, if it's there, then we update the internal state and return true so `std:.remove_if` does its magic.
(bolded the important part)

So, if we remove coins as we find them on the `CoinsResult` vectors from the to_remove set, we will continually search inside a container with less elements up until there is none and we can stop the search earlier.

Reply:
I don't think that's really a useful optimization since `outs` is likely to be very small. If we are concerned about the complexity of `find`, then I think it would be better to use an `unordered_set` rather than erasing after each removal.

This isn't really a blocking comment, I just found it a bit confusing.

Reply:
yeah, an `unordered_set` sounds good too.

still, what I wanted to add (which I actually didn't add it for some reason..) is the capability of stopping early if the set has no elements. E.g. if the wallet has 100k UTXO inside, the preset inputs set is small, and the coins appear early in the loop, the process could stop right away (no need to keep traversing the entire available coins vector if we already found them).